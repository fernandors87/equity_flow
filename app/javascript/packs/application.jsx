import React from 'react'
import ReactDOM from 'react-dom'

import moment from 'moment'
import {Set} from 'immutable'

import * as API from './../src/api'
import '../src/application.scss'
import TransactionSummary from './transaction_summary'

class Application extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      accounts: Set(),
      splits: Set()
    }
  }

  componentDidMount() {
    const accounts = API.accounts.list()
    const splits = API.splits.list()
    Promise.all([accounts, splits]).then(([accounts, splits]) => {
      this.setState({accounts, splits})
    })
  }

  render () {
    return (
      <div>
        <TransactionSummary
          accounts={this.state.accounts}
          splits={this.state.splits}
          startDate={moment("2017-07-01").utc()}
          endDate={moment("2017-12-31").utc()}
          />
      </div>
    )
  }
}

window.api = API

document.addEventListener('DOMContentLoaded', () => {
  ReactDOM.render(
    <Application />,
    document.body.appendChild(document.createElement('div')),
  )
})
