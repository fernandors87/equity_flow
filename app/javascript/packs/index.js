import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import Application from 'components/application'
import Chrono from 'models/chrono'
import { List } from 'immutable'
import { Provider } from 'react-redux'
import React from 'react'
import ReactDOM from 'react-dom'
import accountsReducer from 'reducers/accounts-reducer'
import chronoReducer from 'reducers/chrono-reducer'
import moment from 'moment'
import splitsReducer from 'reducers/splits-reducer'
import thunk from 'redux-thunk'

const reducers = combineReducers({
  accounts: accountsReducer,
  chrono: chronoReducer,
  splits: splitsReducer,
})

const states = {
  accounts: List(),
  chrono: new Chrono(moment().subtract(1, 'year'), moment()),
  splits: List(),
}

const storeEnhancers = compose(
  applyMiddleware(thunk),
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

const store = createStore( reducers, states, storeEnhancers)

document.addEventListener('DOMContentLoaded', () => {
  ReactDOM.render(
    <Provider store={ store }><Application /></Provider>,
    document.body.appendChild(document.createElement('div'))
  )
})
