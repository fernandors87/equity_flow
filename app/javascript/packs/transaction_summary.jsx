import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import * as _ from 'lodash'
import { Map } from 'immutable'

export default class TransactionSummary extends React.Component {

  constructor(props) {
    super(props)
  }

  months(startDate, endDate) {
    if (startDate > endDate) {
      throw `startDate(${startDate}) must not be greater than endDate(${endDate})`
    }

    const startOfMonth = moment(startDate).utc().startOf('month')
    const endOfMonth = moment(endDate).utc().endOf('month')

    const calc = (acc) => {
      const lastProcessed = _.first(acc)
      const plusOne = moment(lastProcessed).add(1, 'month')
      return plusOne.isBefore(endOfMonth) ? calc(_.concat(plusOne, acc)) : acc
    }
    const dates = calc([startOfMonth])
    return dates.map(m => m.toDate()).reverse()
  }

  categories(accounts, months) {
    const monthsMap = months.reduce((acc, month) => {
      const key = moment(month).utc().format('YYYY-MM-DD')
      return acc.set(key, 0)
    }, Map())

    return accounts.reduce((acc, account) => acc.set(account.id, monthsMap), Map())
  }

  cells(categories, transactions) {
    return transactions.reduce((acct, tx) => {
      const month = moment(tx.date).utc().startOf('month').format('YYYY-MM-DD')
      return tx.splits.reduce((accs, s) => {
        const value = s.position == "credit" ? -s.value : s.value
        const monthsMap = accs.get(s.account_id) || Map()
        const currentValue = monthsMap.get(month) || 0
        const updatedMonthsMap = monthsMap.set(month, currentValue + value)
        return accs.set(s.account_id, updatedMonthsMap)
      }, acct)
    }, categories)
  }

  render() {
    const months = this.months(this.props.startDate, this.props.endDate)
    const headers = months.map(d => {
      const month = moment(d).utc().format('YYYY-MM')
      return <th key={month}>{month}</th>
    })

    const categories = this.categories(this.props.accounts, months)
    const cells = this.cells(categories, this.props.transactions)
    const body = cells.map((row, account_id) => {
      const account = this.props.accounts.find(a => a.id == account_id) || {}
      const tableCells = row.map((value, month) => {
        const key = [account_id, month].join()
        return <td key={key}>{value}</td>
      }).toArray()

      return (
        <tr key={account_id}>
          <td>{account.name}</td>
          {tableCells}
        </tr>
      )
    }).toArray()

    return (
      <div>
        <table class="table">
          <thead>
            <tr>
              <th>Account</th>
              {headers}
            </tr>
          </thead>
          <tbody>
            {body}
          </tbody>
        </table>
      </div>
    )
  }
}

TransactionSummary.defaultProps = {
  accounts: [],
  transactions: []
};
