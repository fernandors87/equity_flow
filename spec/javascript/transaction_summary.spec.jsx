import React from 'react'
import TransactionSummary from 'packs/transaction_summary'
import { months, sortedAccounts, accountsTree } from 'packs/transaction_summary'
import renderer from 'react-test-renderer'
import { Map, List, OrderedSet, Set } from 'immutable'
import moment from 'moment'

describe('TransactionSummary', () => {

  const accounts = Set([
    { id: 1, name: 'account 1', level: 0, parent_id: null },
    { id: 2, name: 'account 2', level: 1, parent_id: 1 }
  ])

  const splits = Set([
    {
      id: 1,
      account_id: 3,
      deal_id: 1,
      date: moment('2018-01-20').utc(),
      value: -9.99
    },
    {
      id: 2,
      account_id: 1,
      deal_id: 2,
      date: moment('2018-01-01').utc(),
      value: -9.99
    },
    {
      id: 3,
      account_id: 3,
      deal_id: 2,
      date: moment('2018-01-01').utc(),
      value: -9.99
    },
    {
      id: 4,
      account_id: 1,
      deal_id: 3,
      date: moment('2018-02-20').utc(),
      value: 9.99
    },
    {
      id: 5,
      account_id: 2,
      deal_id: 3,
      date: moment('2018-02-20').utc(),
      value: -2.5
    },
    {
      id: 7,
      account_id: 2,
      deal_id: 3,
      date: moment('2018-02-21').utc(),
      value: -2.5
    },
    {
      id: 7,
      account_id: 2,
      deal_id: 3,
      date: moment('2018-02-20').utc(),
      value: -4.99
    }
  ])

  describe('months', () => {
    it('generate an array with months covering two dates', () => {
      const actual = months(moment.utc('2017-01-30'), moment.utc('2017-04-10')).map(d => d.unix())
      const expected = List(['2017-01-01', '2017-02-01', '2017-03-01', '2017-04-01']).map(d => moment.utc(d).unix())
      expect(actual).toEqual(expected)
    })

    it('generate an empty array when endDate > startDate', () => {
      const actual = () => months(new Date('2018-01-30'), new Date('2017-04-10'))
      expect(actual).toThrow()
    })
  })

  describe('accountsTree', () => {

    it('generate an account tree sorted by name', () => {
      const accounts = Set([
        { id: 1, name: 'account 6', level: 0, parent_id: null },
        { id: 2, name: 'account 5', level: 0, parent_id: null },
        { id: 3, name: 'account 4', level: 1, parent_id: 1 },
        { id: 4, name: 'account 3', level: 1, parent_id: 1 },
        { id: 5, name: 'account 2', level: 1, parent_id: 2 },
        { id: 6, name: 'account 1', level: 2, parent_id: 3 },
      ])

      const expected = Set([
        {
          id: 2,
          name: 'account 5',
          level: 0,
          parent_id: null,
          children: Set([
            {
              id: 5,
              name: 'account 2',
              level: 1,
              parent_id: 2,
              children: Set()
            }
          ])
        },
        {
          id: 1,
          name: 'account 6',
          level: 0,
          parent_id: null,
          children: Set([
            {
              id: 4,
              name: 'account 3',
              level: 1,
              parent_id: 1,
              children: Set()
            },
            {
              id: 3,
              name: 'account 4',
              level: 1,
              parent_id: 1,
              children: Set([
                {
                  id: 6,
                  name: 'account 1',
                  level: 2,
                  parent_id: 3,
                  children: Set()
                }
              ])
            }
          ])
        }
      ])

      const actual = accountsTree(accounts)
      expect(actual).toEqual(expected)
    })
  })

  it('should render a table with monthly account values', () => {
    const component = renderer.create(
      <TransactionSummary
        accounts={accounts}
        splits={splits}
        startDate={moment('2018-01-10').utc()}
        endDate={moment('2018-03-20').utc()} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render nicely with default props', () => {
    const component = renderer.create(<TransactionSummary />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
