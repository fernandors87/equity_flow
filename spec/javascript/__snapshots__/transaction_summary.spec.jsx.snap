// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionSummary render a table with monthly account values 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          Account
        </th>
        <th>
          2017-01
        </th>
        <th>
          2017-02
        </th>
        <th>
          2017-03
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          account 1
        </td>
        <td>
          1000
        </td>
        <td>
          999.99
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          account 2
        </td>
        <td>
          -1000
        </td>
        <td>
          -99.99
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          account 3
        </td>
        <td>
          0
        </td>
        <td>
          -900
        </td>
        <td>
          0
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
